File: na/Ch3/Programing/src/E1.cpp
#include "CurveFitting.h"
#include <fstream>
#include <iostream>
#include <stdexcept>

void outputFittedData(const std::string& filename, const Spline<3, B_spline>& splineX, const Spline<3, B_spline>& splineY, double step = 0.002) {
    std::ofstream file(filename);
    file << std::fixed << std::setprecision(6);
    for (double t = 0.0; t <= 1.0; t += step) {
        double x = splineX(t);
        double y = splineY(t);
        file << x << "," << y << "\n";
    }
    file.close();
}

void outputHeartCurveResults(int N, const std::string& filename) {
    CurveFitting curveFitter;

    // Generate heart curve points
    std::vector<double> x, y, t;
    curveFitter.generateHeartCurveR1(N, x, y);

    // Calculate chordal length parameterization
    curveFitter.calculateChordLengthParam(x, y, t);

    // Fit cubic B-spline
    Spline<3, B_spline> splineX, splineY;
    curveFitter.fitBSpline2D(t, x, y, splineX, splineY, BCType::natural, BCType::natural);

    // Output fitted data
    outputFittedData(filename, splineX, splineY);
}

void outputHeartCurveResults2(int N, const std::string& filename) {
    CurveFitting curveFitter;

    // Generate heart curve points
    std::vector<double> x, y, t;
    curveFitter.generateHeartCurveR1(N, x, y);

    // Calculate chordal length parameterization
    curveFitter.calculateSameDistanceParam(x, t);

    // Fit cubic B-spline
    Spline<3, B_spline> splineX, splineY;
    curveFitter.fitBSpline2D(t, x, y, splineX, splineY, BCType::natural, BCType::natural);

    // Output fitted data
    outputFittedData(filename, splineX, splineY);
}

int main() {
    try {
        // Generate and output results for N = 10, 40, 160
        outputHeartCurveResults(10, "heart_1_curve_N10.csv");
        outputHeartCurveResults(40, "heart_1_curve_N40.csv");
        outputHeartCurveResults(160, "heart_1_curve_N160.csv");

        outputHeartCurveResults(10, "heart_2_curve_N10.csv");
        outputHeartCurveResults(40, "heart_2_curve_N40.csv");
        outputHeartCurveResults(160, "heart_2_curve_N160.csv");

        std::cout << "Heart curve data written to CSV files." << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: na/Ch3/Programing/src/E1.py
import numpy as np
import matplotlib.pyplot as plt

def load_curve_data(filename):
    """Load curve data from a file."""
    data = np.loadtxt(filename, delimiter=",")
    x, y = data[:, 0], data[:, 1]
    # Append the first point to the end to close the curve
    x = np.append(x, x[0])
    y = np.append(y, y[0])
    return x, y


def plot_heart_curves(N_list, output_filenames, output_image):
    """Plot heart curves for different N values."""
    plt.figure(figsize=(10, 6))

    # Plot fitted cubic splines for each N
    for N, filename in zip(N_list, output_filenames):
        x, y = load_curve_data(filename)
        plt.plot(x, y, label=f"Cubic Spline (N={N})")

    plt.title("Heart Curve Fitting for Different N Values")
    plt.xlabel("x")
    plt.ylabel("y")
    plt.axis("equal")
    plt.legend()
    plt.grid()
    plt.savefig("../pics/E1_" + output_image)

if __name__ == "__main__":
    # N values and corresponding filenames
    N_list = [10, 40, 160]
    output_filenames = ["heart_1_curve_N10.csv", "heart_1_curve_N40.csv", "heart_1_curve_N160.csv"]
    output_image = "heart_curve_comparison.png"

    # Plot the curves
    plot_heart_curves(N_list, output_filenames, output_image)

    output_filenames2 = ["heart_2_curve_N10.csv", "heart_2_curve_N40.csv", "heart_2_curve_N160.csv"]
    output_image2 = "heart_curve_comparison_2.png"

    # Plot the curves
    plot_heart_curves(N_list, output_filenames2, output_image2)
